### ISSUE TITLE
Refactor getCost::BaseCostDecider to reduce CCN

### ISSUE THREAD

**Issue by Kappenn** [johanmoritz/freecol#57]
Before improvment the CCN is 24 according to Lizard
The code is:
```java
    public int getCost(final Unit unit, final Location oldLocation,
                       final Location newLocation, int movesLeftBefore) {
        int cost = 0;
        this.newTurns = 0;
              
        Tile oldTile = oldLocation.getTile();
        Tile newTile = newLocation.getTile();
        if (oldLocation instanceof Europe) {
            CodeCoverage.run("getCost");
            // Coming from Europe
            if (newLocation instanceof Europe
                || newTile == null
                || !newTile.isDirectlyHighSeasConnected()
                || !unit.getType().canMoveToHighSeas()){
                CodeCoverage.run("getCost");
                return ILLEGAL_MOVE;
            }else{
                CodeCoverage.run("getCost");
            }
            newTurns = unit.getSailTurns();
            movesLeft = unit.getInitialMovesLeft();
            cost = newTurns * unit.getInitialMovesLeft();
            CodeCoverage.run("getCost");

        } else if (oldTile == null) {
            CodeCoverage.run("getCost");
            return ILLEGAL_MOVE;

        } else if (newLocation instanceof Europe) { // Going to Europe
            if (!unit.getType().canMoveToHighSeas()) {
                CodeCoverage.run("getCost");
                return ILLEGAL_MOVE;
            }else{
                CodeCoverage.run("getCost");
            }
            newTurns = unit.getSailTurns();
            movesLeft = unit.getInitialMovesLeft();
            cost = newTurns * unit.getInitialMovesLeft();
            CodeCoverage.run("getCost");

        } else if (newTile == null || !newTile.isExplored()) {
            CodeCoverage.run("getCost");
            return ILLEGAL_MOVE;

        } else { // Moving between tiles
            // Disallow illegal moves.
            // Special moves and moving off a carrier consume a whole turn.
            CodeCoverage.run("getCost");
            boolean consumeMove = false;
            switch (unit.getSimpleMoveType(oldTile, newTile)) {
            case MOVE_HIGH_SEAS:
                CodeCoverage.run("getCost");
                break;
            case ATTACK_UNIT:
                CodeCoverage.run("getCost");  
                // Ignore hostile units in the base case, treating attacks
                // as moves.
            case MOVE:
                CodeCoverage.run("getCost");
                if (!unit.isOnCarrier()){
                    CodeCoverage.run("getCost");
                    break; // Fall through if disembarking.
                }
            case ATTACK_SETTLEMENT:
                CodeCoverage.run("getCost");
            case EXPLORE_LOST_CITY_RUMOUR:
                CodeCoverage.run("getCost");
            case EMBARK:
                CodeCoverage.run("getCost");
            case ENTER_INDIAN_SETTLEMENT_WITH_FREE_COLONIST:
                CodeCoverage.run("getCost");
            case ENTER_INDIAN_SETTLEMENT_WITH_SCOUT:
                CodeCoverage.run("getCost");
            case ENTER_INDIAN_SETTLEMENT_WITH_MISSIONARY:
                CodeCoverage.run("getCost");
            case ENTER_FOREIGN_COLONY_WITH_SCOUT:
                CodeCoverage.run("getCost");
            case ENTER_SETTLEMENT_WITH_CARRIER_AND_GOODS:
                CodeCoverage.run("getCost");
                consumeMove = true;
                break;
            default:
                CodeCoverage.run("getCost");
                return ILLEGAL_MOVE;
            }

            cost = adjust(unit, oldTile, newTile, movesLeftBefore);
            if (consumeMove) {
                CodeCoverage.run("getCost");
                cost += this.movesLeft;
                this.movesLeft = 0;
            }else{
                CodeCoverage.run("getCost");
            }
        }
        CodeCoverage.run("getCost");  
        return cost;
    }
```

**Comment by Kappenn** [comment]
Moved out the else statement with the switch cases to its own function in 8bb7dfed15bcaf5cf797467b17e64178f0036853.

**Comment by Kappenn** [comment]
The new CCN according to Lizard is 11, the prior is 24, the improvment is more than 35%

**Comment by rbratfors** [comment]
The new CCN is 7 according to me.

**Comment by Kappenn** [comment]
solved in 8bb7dfe
